Fixes case-intolerance in file names. Restores divergent names in tests.

busykai opened the pull request:
Both systemd-class and shim-systemd are fixed in separate commits.

Addresses the concerns raised in #88.

CC: @bryteise, @ikeydoherty

ikeydoherty commented:
No, no, no. The system was FINE as it was, and 8.3 mapping is an assumption, which does not correlate to longer paths (i.e. `/EFI/org.clearlinux` or `/EFI/com.solus-project`).

Why do you have to keep breaking what works? Just restore the original behaviour.

ikeydoherty commented:
Note this is now a workaround to an existing fix, relying on internal kernel driver details, in an *incomplete* fashion, when all you had to do was revert your change, instead of layering more and more stuff on top of it.

busykai commented:
@ikeydoherty this PR just restores the intentionally incorrect cases in the tests and fixes both `systemd-class` and `shim-systemd` with respect to case handling, as per your initial recommendation.

[The comment to #88](https://github.com/clearlinux/clr-boot-manager/pull/88#issuecomment-341359166) about FAT behaviour and all caps is just an analysis of the issues people were having. I didn't implement and this PR is not an implementation of that option. Changes to the [`systemd-class.c`](https://github.com/clearlinux/clr-boot-manager/pull/89/files#diff-dba59e9ff1e1345490d8c62619c73516) are rather simple. And the harness/tests has all the layout names restored as we agreed. I needed a change in harness to check for default bootloader in image mode.

ikeydoherty commented:
No - you've replaced usage of `nc_build_case_correct_path`. My recommendation was to put the test paths back how they were. The systemd-class bootloader **wasn't broken** - diverging away from a proven solution to something else - it makes no sense.

Personally, I reject this PR because it is out of scope and is building upon a previous failure. We just need the test case names restored, and any file path manipulation on the ESP to use `nc_build_case_correct_path`.

busykai commented:
@ikeydoherty, oh, I see. So I compared v1.5.5 with this branch. Here is the view of [systemd-class.c](https://github.com/clearlinux/clr-boot-manager/compare/81da1324fa0e...3de35c4977b7#diff-dba59e9ff1e1345490d8c62619c73516). I do not see any use of `nc_build_case_correct_path` replaced. For `shim-systemd` this PR actually does employ `nc_build_case_correct_path` for path resolution before doing any copy.

When making this change, I was addressing your concerns 1) do not assume any specific case on FAT partition; 2) test with the harness which intentionally uses case-incorrect names.

ikeydoherty commented on src/bootloaders/shim-systemd.c:
```diff
@@ -214,57 +223,46 @@ static bool shim_systemd_needs_update(__cbm_unused__ const BootManager *manager)
-        path = string_printf("%s%s", boot_root, SYSTEMD_ENTRIES);
-        if (!nc_mkdir_p(path, 00755)) {
+
+        systemd_config_entries = string_printf("%s/" SYSTEMD_CONFIG_DIR "/" SYSTEMD_ENTRIES_DIR, boot_root);
```
Should be using `nc_build_case_correct_path`

ikeydoherty commented on src/bootloaders/shim-systemd.c:
```diff
@@ -214,57 +223,46 @@ static bool shim_systemd_needs_update(__cbm_unused__ const BootManager *manager)
-                path = string_printf("%s%s", boot_root, EFI_FALLBACK_DIR);
-                LOG_INFO("Image mode: creating dir for fallback: %s", path);
-                if (!nc_mkdir_p(path, 00755)) {
+                char *efi_fallback_dir = dirname(strdup(efi_fallback_dst_host));
```
strdup can fail, so can dirname.

busykai replied:
Yes, I'll make `get_kernel_destination()` return full path and get rid of all this ugliness. Sorry.

ikeydoherty commented on src/bootloaders/shim-systemd.c:
```diff
@@ -214,57 +223,46 @@ static bool shim_systemd_needs_update(__cbm_unused__ const BootManager *manager)
-                if (!nc_mkdir_p(path, 00755)) {
+                char *efi_fallback_dir = dirname(strdup(efi_fallback_dst_host));
+                if (!nc_mkdir_p(efi_fallback_dir, 00755)) {
+                        free(efi_fallback_dir);
```
We have autofree system.

ikeydoherty commented on src/bootloaders/shim-systemd.c:
```diff
@@ -371,6 +377,11 @@ static void shim_systemd_destroy(const BootManager *manager)
         free(systemd_src);
```
This is all unnecessary - we have autofree.

ikeydoherty commented on tests/check-uefi.c:
```diff
@@ -263,7 +271,7 @@ START_TEST(bootman_uefi_remove_bootloader)
 
         fail_if(!nc_file_exists(TOP_BUILD_DIR "/tests/update_playground/" BOOT_DIRECTORY
-                                              "/EFI/Boot"),
+                                              "/efi/BOOT"),
```
This is not the original case.

busykai replied:
[this line in v1.5.5](https://github.com/clearlinux/clr-boot-manager/blob/v1.5.5/tests/check-uefi.c#L266)

Maybe I misunderstand, what do you refer to as original case?

ikeydoherty replied:
At this point its really hard to tell - so I apologize if I'm wrong on some of these. My primary concern is restoring sync between the old and new test cases. That and ensuring we're using nc_ case function for all paths on an ESP (even for joining paths! :))

ikeydoherty commented on tests/check-uefi.c:
```diff
@@ -281,19 +289,19 @@ START_TEST(bootman_uefi_remove_bootloader)
         /* Ensure that it is indeed removed. */
         fail_if(nc_file_exists(TOP_BUILD_DIR "/tests/update_playground" BOOT_DIRECTORY
-                                             "/EFI/Boot" DEFAULT_EFI_BLOB),
+                                             "/efi/BOOT" DEFAULT_EFI_BLOB),
```
Not the original case

ikeydoherty commented on tests/check-uefi.c:
```diff
@@ -281,19 +289,19 @@ START_TEST(bootman_uefi_remove_bootloader)
 #if defined(HAVE_SYSTEMD_BOOT)
         fail_if(nc_file_exists(TOP_BUILD_DIR "/tests/update_playground/" BOOT_DIRECTORY
-                                             "/EFI/systemd"),
+                                             "/efi/systemd"),
```
Not the original case

ikeydoherty commented on tests/check-uefi.c:
```diff
@@ -281,19 +289,19 @@ START_TEST(bootman_uefi_remove_bootloader)
 #elif defined(HAVE_GUMMIBOOT)
         fail_if(nc_file_exists(TOP_BUILD_DIR "/tests/update_playground/" BOOT_DIRECTORY
-                                             "/EFI/gummiboot"),
+                                             "/efi/gummiboot"),
```
Not the original case

ikeydoherty commented on tests/check-uefi.c:
```diff
@@ -281,19 +289,19 @@ START_TEST(bootman_uefi_remove_bootloader)
 #else
         fail_if(nc_file_exists(TOP_BUILD_DIR "/tests/update_playground/" BOOT_DIRECTORY
-                                             "/EFI/goofiboot"),
+                                             "/efi/goofiboot"),
```
Not the original case

ikeydoherty commented on tests/harness.c:
```diff
@@ -49,11 +49,11 @@
 #define BOOT_FULL PLAYGROUND_ROOT "/" BOOT_DIRECTORY
 
-#define EFI_START BOOT_FULL "/EFI"
+#define EFI_START BOOT_FULL "/efi"
```
Not the original case

busykai replied:
[This line in v1.5.5](https://github.com/clearlinux/clr-boot-manager/blob/v1.5.5/tests/harness.c#L51). I just restored the original names.

ikeydoherty commented on tests/harness.c:
```diff
@@ -482,7 +482,7 @@ int kernel_installed_files_count(BootManager *manager, PlaygroundKernel *kernel)
         const char *esp_path = manager->bootloader->get_kernel_destination
                                    ? manager->bootloader->get_kernel_destination(manager)
-                                   : "EFI/" KERNEL_NAMESPACE;
+                                   : "efi/" KERNEL_NAMESPACE;
```
Not the original case

ikeydoherty commented on src/bootloaders/shim-systemd.c:
```diff
@@ -111,40 +114,46 @@ __cbm_export__ const BootLoader
 static char *shim_src;
-static char *shim_dst_host; /* as accessible by the CMB for file ops. */
-static char *shim_dst_esp;  /* absolute location of shim on the ESP. */
 static char *systemd_src;
```
All this globally defined stuff violates how the other bootloader systems are implemented, globals are hard to track

ikeydoherty commented on src/bootloaders/shim-systemd.c:
```diff
@@ -348,16 +346,24 @@ static bool shim_systemd_init(const BootManager *manager)
+            nc_build_case_correct_path(boot_root, ESP_EFI, ESP_BOOT, EFI_FALLBACK, NULL);
+        kernel_dst_host =
+            nc_build_case_correct_path(boot_root, ESP_EFI, KERNEL_NAMESPACE, KERNEL_DST_DIR, NULL);
+        kernel_dst_esp = strdup(kernel_dst_host + strlen(boot_root));
```
...?

busykai replied:
oh, yeah, this is very ugly. it's basically to get absolute path on ESP.

i am thinking that the right solution to avoid this would be to return full mounted path and then use it in `kernel.c`. right now, it [constructs its own path](https://github.com/clearlinux/clr-boot-manager/blob/master/src/bootman/kernel.c#L613) while it could use full path if `get_kernel_destination()` returned full path. we will then not need to do this and `kernel.c` constructing its own path. what do you think?

busykai commented:
@ikeydoherty, I am not sure how to address your "Not the original case" comments. I checked the paths were original before submitting the PR. If you meant something different, please let me know.

bryteise commented:
Okay that last commit looks like what we want, could you ammend that into the previous commit and force push though so I can see them with full context easier?

busykai commented:
Done.
